camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelAnnotator.java

```diff
@@ -0,0 +1,220 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.idea.annotator;
+
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import com.intellij.lang.annotation.AnnotationHolder;
+import com.intellij.lang.annotation.Annotator;
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiElement;
+import org.apache.camel.catalog.CamelCatalog;
+import org.apache.camel.catalog.EndpointValidationResult;
+import org.apache.camel.catalog.SimpleValidationResult;
+import org.apache.camel.idea.catalog.CamelCatalogService;
+import org.apache.camel.idea.util.CamelService;
+import org.apache.camel.idea.util.IdeaUtils;
+import org.jetbrains.annotations.NotNull;
+
+import static org.apache.camel.idea.util.IdeaUtils.isEmpty;
+
+
+/**
+ * Validate Camel URI endpoint and simple expression and annotated the specific property
+ * or simple expression to highlight the error on the editor
+ */
+public class CamelAnnotator implements Annotator {
+    @Override
+    public void annotate(@NotNull PsiElement element, @NotNull AnnotationHolder holder) {
+        if (ServiceManager.getService(element.getProject(), CamelService.class).isCamelPresent()) {
+            String fromElement = IdeaUtils.extractTextFromElement(element);
+            if (fromElement != null && !fromElement.isEmpty() && !fromElement.trim().endsWith("&") && !fromElement.trim().endsWith("?")
+                && !fromElement.trim().endsWith("{{") && !fromElement.trim().endsWith("=")
+                && !fromElement.trim().endsWith("${")) {
+                validateEndpointProperties(element, holder, fromElement);
+                //Comment out until we solve the classloader issue.
+                //validateSimpleExpression(element, holder, fromElement);
+            }
+        }
+    }
+
+    /**
+     * Validate endpoint options list aka properties. eg "timer:trigger?delay=1000&bridgeErrorHandler=true"
+     * if the URI is not valid a error annotation is created and highlight the invalid value.
+     */
+    private void validateEndpointProperties(@NotNull PsiElement element, @NotNull AnnotationHolder holder, String fromElement) {
+        if ((fromElement.contains("?") || fromElement.contains("&")) && IdeaUtils.isConsumerEndpoint(element)) {
```

Comment:
The Camel endpoints can be anywhere, not only from a consumer endpoint. For example in a .to("jms:cheese?xxx") code line. So I dont think you should limit to only that. Also an Camel endpoint without ? or & can still be validated, as it may be invalid in the context-path section as well. However they may be harder to markup in the code editor where that error is.